src/bloqade/native/upstream/squin2native.py

```diff
@@ -0,0 +1,128 @@
+from dataclasses import field, dataclass
+
+from kirin import ir, passes, rewrite
+from kirin.dialects import py, func
+from kirin.rewrite.abc import RewriteRule, RewriteResult
+from kirin.passes.callgraph import CallGraphPass, ReplaceMethods
+from kirin.analysis.callgraph import CallGraph
+
+from bloqade.native import kernel, broadcast
+from bloqade.squin.clifford import stmts, dialect as clifford_dialect
+
+
+class GateRule(RewriteRule):
+    SQUIN_MAPPING: dict[type[ir.Statement], tuple[ir.Method, ...]] = {
+        stmts.X: (broadcast.x,),
+        stmts.Y: (broadcast.y,),
+        stmts.Z: (broadcast.z,),
+        stmts.H: (broadcast.h,),
+        stmts.S: (broadcast.s, broadcast.s_adj),
+        stmts.T: (broadcast.t, broadcast.t_adj),
+        stmts.SqrtX: (broadcast.sqrt_x, broadcast.sqrt_x_adj),
+        stmts.SqrtY: (broadcast.sqrt_y, broadcast.sqrt_y_adj),
+        stmts.Rx: (broadcast.rx,),
+        stmts.Ry: (broadcast.ry,),
+        stmts.Rz: (broadcast.rz,),
+        stmts.CX: (broadcast.cx,),
+        stmts.CY: (broadcast.cy,),
+        stmts.CZ: (broadcast.cz,),
+    }
+
+    def rewrite_Statement(self, node: ir.Statement) -> RewriteResult:
+        if (native_methods := self.SQUIN_MAPPING.get(type(node))) is None:
+            return RewriteResult()
+
+        if isinstance(node, stmts.SingleQubitNonHermitianGate):
+            native_method = native_methods[1] if node.adjoint else native_methods[0]
+        else:
+            native_method = native_methods[0]
+
+        # do not rewrite in invoke because callgraph pass will be looking for invoke statements
+        (callee := py.Constant(native_method)).insert_before(node)
+        node.replace_by(func.Call(callee.result, tuple(node.args), kwargs=()))
+
+        return RewriteResult(has_done_something=True)
+
+
+@dataclass
+class UpdateDialectsOnCallGraph(passes.Pass):
+    """Update All dialects on the call graph to a new set of dialects given to this pass.
+
+    Usage:
+        pass_ = UpdateDialectsOnCallGraph(rule=rule, dialects=new_dialects)
+        pass_(some_method)
+
+    Note: This pass does not update the dialects of the input method, but copies
+    all other methods invoked within it before updating their dialects.
+
+    """
+
+    fold_pass: passes.Fold = field(init=False)
+
+    def __post_init__(self):
+        self.fold_pass = passes.Fold(self.dialects, no_raise=self.no_raise)
+
+    def unsafe_run(self, mt: ir.Method) -> RewriteResult:
+        mt_map = {}
+
+        cg = CallGraph(mt)
+
+        all_methods = set(sum(map(tuple, cg.defs.values()), ()))
+        for original_mt in all_methods:
+            if original_mt is mt:
+                new_mt = original_mt
```

Comment:
Technically, I'd expect this pass to also update the dialects of `mt`, but I don't think that is happening here (unless I'm missing something). It doesn't matter here, since you are creating a `mt.similar` in `SquinToNative` before calling this pass. Should we upstream it, however, I'd suggest to change this.

Reply:
yeah as discussed on slack ,Passes to not copy the method but they can (and should in this case) copy data inside the method

Reply:
There is a copy inside the `SquinToNative.emit` method.